python/arcticdb/version_store/library.py

```diff
@@ -288,6 +302,195 @@ class ReadInfoRequest(NamedTuple):
+        """
+        q = QueryBuilder().prepend(self)
+        q._optimisation = self._optimisation
+        return ReadRequest(
+            symbol=self.read_request.symbol,
+            as_of=self.read_request.as_of,
+            date_range=self.read_request.date_range,
+            row_range=self.read_request.row_range,
+            columns=self.read_request.columns,
+            query_builder=q,
+        )
+
+    def collect(self) -> VersionedItem:
+        """
+        Read the data and execute any queries applied to this object since the read call.
+
+        Returns
+        -------
+        VersionedItem
+            Object that contains a .data and .metadata element.
+        """
+        return self.lib.read(**self.to_read_request()._asdict())
+
+    def __str__(self) -> str:
+        query_builder_repr = super().__str__()
+        return "LazyDataFrame(" + self.read_request.__repr__() + (" | " if len(query_builder_repr) else "") + query_builder_repr + ")"
+
+    # Needs to be explicitly defined for lists of these objects in LazyDataFrameCollection to render correctly
+    def __repr__(self) -> str:
+        return self.__str__()
+
+
+class LazyDataFrameCollection(QueryBuilder):
+    """
+    Lazy dataframe implementation for batch operations. Allows the application of chains of queries to be added before
+    the actual reads are performed. Queries applied to this object will be applied to all  the symbols being read.
+    If per-symbol queries are required, split can be used to break this class into a list of LazyDataFrame objects.
+    Returned by `Library.read_batch` calls when `lazy=True`.
+
+    See Also
+    --------
+    QueryBuilder for supported querying operations.
+
+    Examples
+    --------
+
+    # Specify that we want the latest version of "test_0" symbol, and version 0 of "test_1" symbol
+    >>> lazy_dfs = lib.read_batch(["test_0", ReadRequest("test_1", as_of=0)], lazy=True)
+    # Perform a filtering operation on both the "test_0" and "test_1" symbols
+    >>> lazy_dfs = lazy_dfs[lazy_dfs["col1"].isin(0, 3, 6, 9)]
+    # Perform a different projection operation on each symbol
+    >>> lazy_dfs = lazy_dfs.split()
+    >>> lazy_dfs[0].apply("new_col", lazy_dfs[0]["col1"] + 1)
+    >>> lazy_dfs[1].apply("new_col", lazy_dfs[1]["col1"] + 2)
+    # Bring together again and perform the same filter on both symbols
+    >>> lazy_dfs = LazyDataFrameCollection(lazy_dfs)
+    >>> lazy_dfs = lazy_dfs[lazy_dfs["new_col"] > 0]
+    # Actual read and processing happens here
+    >>> res = lazy_dfs.collect()
+    """
+    def __init__(
+            self,
+            lazy_dataframes: List[LazyDataFrame],
+    ):
+        lib_set = set([lazy_dataframe.lib for lazy_dataframe in lazy_dataframes])
+        check(
+            len(lib_set) in [0, 1],
+            f"LazyDataFrameCollection init requires all provided lazy dataframes to be referring to the same library, but received:\n{[lib for lib in lib_set]}"
```

Comment:
Remove the newline? (makes grepping logs hard)

Reply:
Done